pilot/pkg/networking/core/v1alpha3/httproute.go

```diff
@@ -220,8 +220,16 @@ func (configgen *ConfigGeneratorImpl) buildSidecarOutboundHTTPRouteConfig(
 // selectVirtualServices selects the virtual services by matching given services' host names.
 func selectVirtualServices(virtualServices []config.Config, servicesByName map[host.Name]*model.Service) []config.Config {
 	out := make([]config.Config, 0)
-	for _, c := range virtualServices {
-		rule := c.Spec.(*networking.VirtualService)
+	// Pre-compute wildcarded service hosts to reduce loop count.
+	wildCardedSvcHosts := []host.Name{}
+	for svcHost := range servicesByName {
+		if svcHost.IsWildCarded() {
+			wildCardedSvcHosts = append(wildCardedSvcHosts, svcHost)
```

Comment:
I think this applies to very small case, for most k8s services, they are not wildcarded host name.

Reply:
Yes. I want to split it from fqdn hosts.

In my environment, loop between fqdn vs hosts and fqdn svc hosts will consume too much cpu.

Reply:
On the contrary, the optimization effect of this change will be very obvious when there are few wildcard hosts.

For example

```
vs hosts:
a.com, nomatch1.com, nomatch2.com, nomatch3.com

svc hosts:
a.com, b.com, c.com, d.com
```

Loop count before this change:
```
a.com 1
nomatch1.com 4
nomatch2.com 4
nomatch3.com 4
```

Each nomatch host will iterate over all svc hosts(actually only wildcard hosts are needed because fqdn hosts are excluded by map searching).

Loop count after this change:
```
a.com 1
nomatch1.com 1
nomatch2.com 1
nomatch3.com 1
```

Reply:
I am also confused, not understand what you do

Reply:
@hzxuzhonghu This case:

```
vs hosts:
a.com, nomatch1.example.com, nomatch2.com, nomatch3.com

svc hosts:
a.com, b.com, c.com, d.com, *.example.com
```

Focus on the `nomatch1.com` loop

```go
for _, h := range rule.Hosts {
	// h: nomatch1.example.com
	// not in svc hosts map [a.com, b.com, c.com, d.com, *.example.com]
	if servicesByName[host.Name(h)] != nil {
		match = true
		break
	}

        // Continue iterating over all the svc hosts [a.com, b.com, c.com, d.com, *.example.com]
        // But [a.com, b.com, c.com, d.com] are excluded by map searching above.
        // Only need to iterate over [*.example.com] and will match.
	for svcHost := range servicesByName {
		if host.Name(h).Matches(svcHost) {
			match = true
			break
		}
	}

	if match {
		break
	}
}
```

I'm not sure if I miss something.

Reply:
kind of understand, please add some comments to the code

Reply:
this can remove {both not wildcard host - host} match  again

Reply:
Oh its for optimizing the "no match" case. Got it. Thanks

Reply:
Yeah. We already do similar optimization else where https://github.com/istio/istio/blob/36f988f68a40fb9aecbf84e0004805770fa9afaa/pilot/pkg/networking/core/v1alpha3/route/route.go#L171 also

Reply:
Comments added.

Reply:
@hzxuzhonghu My mistake. I don't check in the file changes.

Create a new PR https://github.com/istio/istio/pull/39294